File: demos/UnityModuleTest/Packages/Xrpa/Runtime/utils/XrpaTypes.cs
/*
// @generated
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


using System;
using System.Diagnostics;

namespace Xrpa
{

    public abstract class ObjectAccessorInterface
    {
        protected MemoryAccessor _memAccessor;

        public void SetAccessor(MemoryAccessor memAccessor)
        {
            _memAccessor = memAccessor;
        }

        public bool IsNull()
        {
            return _memAccessor == null || _memAccessor.IsNull();
        }
    }

    public struct HashValue
    {
        public static readonly int DS_HASH_SIZE = 32;

        public ulong Value0;
        public ulong Value1;
        public ulong Value2;
        public ulong Value3;

        public HashValue(ulong part0, ulong part1, ulong part2, ulong part3)
        {
            Value0 = part0;
            Value1 = part1;
            Value2 = part2;
            Value3 = part3;
        }

        public static bool operator ==(HashValue left, HashValue right)
        {
            return left.Value0 == right.Value0 && left.Value1 == right.Value1 && left.Value2 == right.Value2 && left.Value3 == right.Value3;
        }

        public static bool operator !=(HashValue left, HashValue right)
        {
            return left.Value0 != right.Value0 || left.Value1 != right.Value1 || left.Value2 != right.Value2 || left.Value3 != right.Value3;
        }

        public bool Equals(HashValue other)
        {
            if (ReferenceEquals(this, other))
            {
                return true;
            }
            return this == other;
        }

        public override bool Equals(object obj)
        {
            if (obj is HashValue identifier)
            {
                return Equals(identifier);
            }
            return false;
        }

        public override int GetHashCode()
        {
            return System.HashCode.Combine(Value0, Value1, Value2, Value3);
        }

        public static HashValue ReadValue(MemoryAccessor memAccessor, MemoryOffset offset)
        {
            var v0 = memAccessor.ReadUlong(offset);
            var v1 = memAccessor.ReadUlong(offset);
            var v2 = memAccessor.ReadUlong(offset);
            var v3 = memAccessor.ReadUlong(offset);
            return new HashValue
            {
                Value0 = v0,
                Value1 = v1,
                Value2 = v2,
                Value3 = v3
            };
        }

        public static void WriteValue(HashValue value, MemoryAccessor memAccessor, MemoryOffset offset)
        {
            memAccessor.WriteUlong(value.Value0, offset);
            memAccessor.WriteUlong(value.Value1, offset);
            memAccessor.WriteUlong(value.Value2, offset);
            memAccessor.WriteUlong(value.Value3, offset);
        }
    }

    public class TransportConfig
    {
        public HashValue SchemaHash;
        public int ChangelogByteCount;
    }

    public struct ObjectUuid : System.IEquatable<ObjectUuid>
    {
        public ulong ID0;
        public ulong ID1;

        public ObjectUuid(ulong part0, ulong part1)
        {
            ID0 = part0;
            ID1 = part1;
        }

        public ObjectUuid(uint A, uint B, uint C, uint D)
        {
            ID0 = (((ulong)A) << 32) | B;
            ID1 = (((ulong)C) << 32) | D;
        }

        public ObjectUuid(System.Guid guid)
        {
            var bytes = guid.ToByteArray();
            uint A = ((uint)bytes[0]) << 24 | ((uint)bytes[1]) << 16 | ((uint)bytes[2]) << 8 | ((uint)bytes[3]);
            uint B = ((uint)bytes[4]) << 24 | ((uint)bytes[5]) << 16 | ((uint)bytes[6]) << 8 | ((uint)bytes[7]);
            uint C = ((uint)bytes[8]) << 24 | ((uint)bytes[9]) << 16 | ((uint)bytes[10]) << 8 | ((uint)bytes[11]);
            uint D = ((uint)bytes[12]) << 24 | ((uint)bytes[13]) << 16 | ((uint)bytes[14]) << 8 | ((uint)bytes[15]);
            ID0 = (((ulong)A) << 32) | B;
            ID1 = (((ulong)C) << 32) | D;
        }

        public void Clear()
        {
            ID0 = 0;
            ID1 = 0;
        }

        public bool IsEmpty()
        {
            return ID0 == 0 && ID1 == 0;
        }

        public static bool operator ==(ObjectUuid left, ObjectUuid right)
        {
            return left.ID0 == right.ID0 && left.ID1 == right.ID1;
        }

        public static bool operator !=(ObjectUuid left, ObjectUuid right)
        {
            return left.ID0 != right.ID0 || left.ID1 != right.ID1;
        }

        public bool Equals(ObjectUuid other)
        {
            if (ReferenceEquals(this, other))
            {
                return true;
            }
            return ID0 == other.ID0 && ID1 == other.ID1;
        }

        public override bool Equals(object obj)
        {
            if (obj is ObjectUuid identifier)
            {
                return Equals(identifier);
            }
            return false;
        }

        public override int GetHashCode()
        {
            return System.HashCode.Combine(ID0, ID1);
        }

        public int Compare(ObjectUuid other)
        {
            if (ID0 == other.ID0)
            {
                if (ID1 == other.ID1)
                {
                    return 0;
                }
                return ID1 < other.ID1 ? -1 : 1;
            }
            return ID0 < other.ID0 ? -1 : 1;
        }

        public static ObjectUuid ReadValue(MemoryAccessor memAccessor, MemoryOffset offset)
        {
            var id0 = memAccessor.ReadUlong(offset);
            var id1 = memAccessor.ReadUlong(offset);
            return new ObjectUuid(id0, id1);
        }

        public static void WriteValue(ObjectUuid value, MemoryAccessor memAccessor, MemoryOffset offset)
        {
            memAccessor.WriteUlong(value.ID0, offset);
            memAccessor.WriteUlong(value.ID1, offset);
        }
    }

}

File: modules/xred-debug/ImageViewer/xrpa_runtime/utils/xrpa_types.py
# @generated
# (c) Meta Platforms, Inc. and affiliates. Confidential and proprietary.


import uuid
from dataclasses import dataclass

from xrpa_runtime.utils.memory_accessor import MemoryAccessor, MemoryOffset


class ObjectAccessorInterface:
    def __init__(self, mem_accessor: "MemoryAccessor" = None):
        self._mem_accessor = mem_accessor

    def is_null(self) -> bool:
        return self._mem_accessor is None or self._mem_accessor.is_null()


class HashValue:
    def __init__(self, part0: int, part1: int, part2: int, part3: int):
        self.Value0 = part0
        self.Value1 = part1
        self.Value2 = part2
        self.Value3 = part3

    def __eq__(self, other: "HashValue"):
        return (
            self.Value0 == other.Value0
            and self.Value1 == other.Value1
            and self.Value2 == other.Value2
            and self.Value3 == other.Value3
        )

    def __ne__(self, other: "HashValue"):
        return not self.__eq__(other)

    def __hash__(self):
        return hash((self.Value0, self.Value1, self.Value2, self.Value3))

    @staticmethod
    def read_value(mem_accessor, offset: MemoryOffset) -> "HashValue":
        v0 = mem_accessor.read_ulong(offset)
        v1 = mem_accessor.read_ulong(offset)
        v2 = mem_accessor.read_ulong(offset)
        v3 = mem_accessor.read_ulong(offset)
        return HashValue(v0, v1, v2, v3)

    @staticmethod
    def write_value(value, mem_accessor, offset: MemoryOffset):
        mem_accessor.write_ulong(value.Value0, offset)
        mem_accessor.write_ulong(value.Value1, offset)
        mem_accessor.write_ulong(value.Value2, offset)
        mem_accessor.write_ulong(value.Value3, offset)


@dataclass(frozen=True)
class TransportConfig:
    schema_hash: HashValue
    changelog_byte_count: int


class ObjectUuid:
    def __init__(self, part0: int, part1: int):
        self.ID0 = part0
        self.ID1 = part1

    @classmethod
    def from_uint(cls, A: int, B: int, C: int, D: int):
        ID0 = ((A) << 32) | B
        ID1 = ((C) << 32) | D
        return cls(ID0, ID1)

    @classmethod
    def from_uuid(cls, guid: uuid.UUID):
        bytes = guid.bytes
        A = (bytes[0]) << 24 | (bytes[1]) << 16 | (bytes[2]) << 8 | (bytes[3])
        B = (bytes[4]) << 24 | (bytes[5]) << 16 | (bytes[6]) << 8 | (bytes[7])
        C = (bytes[8]) << 24 | (bytes[9]) << 16 | (bytes[10]) << 8 | (bytes[11])
        D = (bytes[12]) << 24 | (bytes[13]) << 16 | (bytes[14]) << 8 | (bytes[15])
        return cls.from_uint(A, B, C, D)

    def clear(self):
        self.ID0 = 0
        self.ID1 = 0

    def is_empty(self) -> bool:
        return self.ID0 == 0 and self.ID1 == 0

    def __eq__(self, other: "ObjectUuid"):
        return self.ID0 == other.ID0 and self.ID1 == other.ID1

    def __ne__(self, other: "ObjectUuid"):
        return not self.__eq__(other)

    def __hash__(self: "ObjectUuid"):
        return hash((self.ID0, self.ID1))

    def compare(self, other: "ObjectUuid"):
        if self.ID0 == other.ID0:
            if self.ID1 == other.ID1:
                return 0
            return self.ID1 < other.ID1 and -1 or 1
        return self.ID0 < other.ID0 and -1 or 1

    @staticmethod
    def read_value(mem_accessor: MemoryAccessor, offset: MemoryOffset) -> "ObjectUuid":
        id0 = mem_accessor.read_ulong(offset)
        id1 = mem_accessor.read_ulong(offset)
        return ObjectUuid(id0, id1)

    @staticmethod
    def write_value(
        value: "ObjectUuid", mem_accessor: MemoryAccessor, offset: MemoryOffset
    ):
        mem_accessor.write_ulong(value.ID0, offset)
        mem_accessor.write_ulong(value.ID1, offset)
